### PR TITLE
add env vars to ecs run launcher

### LINKED ISSUE #7282: Allow environment variables in ECS Run Launcher
## Use Case

Allow the ECS run launcher to except [environment variables](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/taskdef-envfiles.html) to supply to the launched task rather. Currently only secrets can be supplied.

## Ideas of Implementation

Add an additional input parameter to the `EcsRunLauncher` to all for a mapping of environment variables.

## Additional Info

Use case is for none sensitive environment variables to be set that do not need to be stored in secrets.

---
#### Message from the maintainers:
Excited about this feature? Give it a :thumbsup:. We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by jamiedemaria**
### Summary & Motivation
resolves #7282

### How I Tested These Changes
bk

### CODE REVIEW

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py [hunk 1]
```diff
@@ -70,3 +79,7 @@ def test_merge(
         empty_container_context.merge(secrets_container_context).secrets_tags
         == secrets_container_context.secrets_tags
     )
+
+    merged = merged.merge(environment_container_context)
```
Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py:83]: This isn't really testing any merge behaviors. I'd probably modify the existing container context fixtures to include environment. And then assert things like:

- what happens if I merge a container context with environment variable `FOO` into a container context without that environment variable?
- what happens if I merge a container context with environment variable `FOO` into a container context without any environment variables?
- what happens if I merge a container context with environment variable `FOO` into a container context that already has environment variable `FOO`? Which one takes precedence?
- what happens if I merge a container context without environment variable `FOO` into a container context with environment variable `FOO`?

Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py:83]: This isn't really testing any merge behaviors. I'd probably modify the existing container context fixtures to include environment. And then assert things like:

- what happens if I merge a container context with environment variable `FOO` into a container context without that environment variable?
- what happens if I merge a container context with environment variable `FOO` into a container context without any environment variables?
- what happens if I merge a container context with environment variable `FOO` into a container context that already has environment variable `FOO`? Which one takes precedence?
- what happens if I merge a container context without environment variable `FOO` into a container context with environment variable `FOO`?

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+def test_environment_with_container_context(
```
Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py:1]: Can you also add a test for setting the environment via instance config? Similar to https://github.com/dagster-io/dagster/blob/f305b65e057eb73c6be70ebb430c608941a62e40/python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_secrets.py#L9

Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py:1]: Can you also add a test for setting the environment via dagster.yaml?

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    # It includes environment vars
+    environment = task_definition["containerDefinitions"][0]["environment"]
+    assert environment[0] == {
+        "name": config["environment"][0]["name"],
+        "value": config["environment"][0]["value"],
+    }
+
+    # But no other env vars
+    assert len(environment) == 1
```
Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py:31]: I don't think this assertion is particularly important in the context of environment.

For secrets, we're pulling existing secrets from AWS Secrets Manager or AWS SSM Parameter Store. So we have some assertions to make sure we're only pulling a subset of those secrets.

Environment variables are a bit different - there's no canonical source of them. I wouldn't expect to see any environment variable that we aren't explicitly defining somewhere (either via instance config or via container context).

Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py:31]: I don't think this assertion is particularly important in the context of environment.

For secrets, we're pulling existing secrets from AWS Secrets Manager or AWS SSM Parameter Store. So we have some assertions to make sure we're only pulling a subset of those secrets.

Environment variables are a bit different - there's no canonical source of them. I wouldn't expect to see any environment variable that we aren't explicitly defining somewhere (either via instance config or via container context).

**Review by jmsanders — DISMISSED** [review]
I think the implementation looks pretty good but it's a bit hard to know for sure because there are some test cases missing. Can you test a few more scenarios to increase confidence?

Also, any plans to support environment files? I noticed them while reading through the docs you linked in a comment. Perhaps not part of this PR, but something for us to consider in the future. https://docs.aws.amazon.com/AmazonECS/latest/developerguide/taskdef-envfiles.html#taskdef-envfiles-considerations

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py (hunk 1 above)
Comment (jmsanders) [reply python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py:101]: Looks like an edge case that wasn't tested for when container context was added for secrets. @gibsondan is this behavior intentional?

I agree that we probably want to change the behavior (and perhaps to choose the "newer" (second) one if there is a collision - whichever direction we go, we should include the test you wrote so we can detect regressions if somebody were to change that behavior in the future.

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py [hunk 4]
```diff
@@ -70,3 +97,20 @@ def test_merge(
...
+        == environment_container_context.environment + secrets_container_context.environment
+    )
+
+    merged = merged.merge(other_environment_container_context)
+
+    assert len(merged.environment) == 4
+    assert merged.environment == [
+        {"name": "Hello", "value": "Goodbye"},
+        {"name": "FOO", "value": "BAZ"},
+        {"name": "FOO", "value": "BAR"},
```
Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py:114]: Yeah, I agree. This is not what I would expect when merging. I don't really know if ECS will resolve that for us or if it'll error 🤷

**Review by jmsanders — APPROVED** [review]

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py (hunk 2 above)
Comment (jmsanders) [reply python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py]: Something like this: https://github.com/dagster-io/dagster/blob/8455843d675eb8f61028cb1e6e2990a8136f0359/python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_secrets.py#L20-L29

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py (hunk 1 above)
Comment (gibsondan) [reply python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_container_context.py:101]: in general when there's a collision the intention is that the other one being passed in wins, yeah. I'd be down for making that improvement here

**Review by gibsondan — COMMENTED** [review]
this is great - we should get these in the container context schema in cloud too so that people can set them right from the workspace tab

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py (hunk 2 above)
Comment (jamiedemaria) [reply python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_environment.py]: ok cool did that in another test in the file

**Pull request closed without merging**